File: python/113_Path_Sum_II.py
#!/usr/bin/env python3
import TreeMethod


class Solution(object):
    def pathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: List[List[int]]
        """
        if root is None:
            return []
        result = self.helper(root, sum, [])
        return result if result is not None else []

    def helper(self, root, sum, curList):
        if root is None:
            if not sum:
                return [curList]
            else:
                return None

        if root.left is None and root.right is not None:
            curList.append(root.val)
            return self.helper(root.right, sum - root.val, curList)
        elif root.right is None and root.left is not None:
            curList.append(root.val)
            return self.helper(root.left, sum - root.val, curList)

        if root.left is None and root.right is None:
            if sum - root.val == 0:
                curList.append(root.val)
                return [curList]
            else:
                return None

        tmpList = curList[:]
        curList.append(root.val)
        tmpList.append(root.val)
        result1 = self.helper(root.left, sum - root.val, curList)
        result2 = self.helper(root.right, sum - root.val, tmpList)
        if result1 is None and result2 is None:
            return None
        elif result1 is not None and result2 is not None:
            result1.extend(result2)
            return result1
        elif result1 is not None:
            return result1
        else:
            return result2





root = TreeMethod.build([5,4,8,11,None,13,4,7,2,None,None,5,1])
s = Solution()
print(s.pathSum(root, 22))

File: java/src/brand_new/_113_Path_Sum_II.java
package brand_new;

import java.util.ArrayList;
import java.util.List;

public class _113_Path_Sum_II {
	public List<List<Integer>> pathSum(TreeNode root, int sum) {
		List<List<Integer>> result = new ArrayList<>();
		pathSum(result, root, sum, new ArrayList<Integer>());
		return result;
	}

	public void pathSum(List<List<Integer>> result, TreeNode root, int sum,
			List<Integer> prefix) {
		if (root == null)
			return;
		if (root.left == null && root.right == null && sum == root.val) {
			prefix.add(root.val);
			result.add(new ArrayList<Integer>(prefix));
			prefix.remove(prefix.size() - 1);
		} else {
			prefix.add(root.val);
			if (root.left != null)
				pathSum(result, root.left, sum - root.val, prefix);
			if (root.right != null)
				pathSum(result, root.right, sum - root.val, prefix);

			prefix.remove(prefix.size() - 1);
		}
	}
}
